core/src/main/java/org/bitcoinj/core/TransactionConfidence.java

```diff
@@ -449,7 +442,9 @@ public TransactionConfidence duplicate() {
      * explicitly, more precise control is available. Note that this will run the listeners on the user code thread.
      */
     public void queueListeners(final Listener.ChangeReason reason) {
-        for (final ListenerRegistration<Listener> registration : listeners) {
+        //Replaced with for loop as iterator init consumpted 11% of cpu time
```

Comment:
This is very unbelievable to me. I suspect your measurement to be wrong as iterators on lists are basically as fast as it can get. I did some [very basic simulation](https://gist.github.com/Giszmo/27d484cb02ef759de51d8beeebedb12d) with the actual classes involved. and with a list of 10k, measurement is single digit. With 1M, setup takes forever. With 300k, setup takes 21s and iterations both less than 10ms.

Now this might be a special issue with Android or something but I would like to rule out false measurement before adding such a degradation. After all, it's even by your measure only 12% compared to some cases where we see 97% of potential.

Reply:
The problem is that assumption that there's a lot of listeners is totally wrong. Number of listeners never exceeds two, so we would never loose in using basic loop, while as this method is called *very* frequently we would loose on object construction.

Reply:
I did more tests, trying to provoke many runs of a short list but I see less than a 10% difference there. If so much time is spent in this part, maybe we can improve it with some kind of batching, as apparently the listeners mostly ignore what they get.